lib/services/publish-service.ts

```diff
@@ -0,0 +1,226 @@
+﻿///<reference path="../.d.ts"/>
+"use strict";
+
+import options = require("../common/options");
+import os = require("os");
+import path = require("path");
+import util = require("util");
+var Table = require("cli-table");
+
+export class PublishService implements IPublishService {
+    private static JSON_PUBLISH_FILE_NAME = ".abpublish";
+    private static ALLOWED_CONNECTION_TYPE = "ftp";
+    private static ERROR_MESSAGE_WHEN_PROJECT_EXISTS_ON_REMOTE = "User cancelled operation";
+    private static PUBLIC_URL_DEFAULT_VALUE = "";
+    private static REMOVE_ALL_CONNECTIONS_MESSAGE = "All of them";
+
+    public allPublishConnections: IPublishConnection[];
+    public ftpPublishConnections: IPublishConnection[];
+
+    constructor(private $errors: IErrors,
+                private $fs: IFileSystem,
+                private $logger: ILogger,
+                private $progressIndicator: IProgressIndicator,
+                private $projectConstants: Project.IProjectConstants,
+                private $projectFilesManager: Project.IProjectFilesManager,
+                private $project: Project.IProject,
+                private $prompter: IPrompter,
+                private $server: Server.IServer) {
+
+        this.$project.ensureProject();
+
+        if (!this.$project.capabilities.publish) {
+            this.$errors.failWithoutHelp("The publish command is only applicable to projects of type %s", this.$projectConstants.TARGET_FRAMEWORK_IDENTIFIERS.MobileWebsite);
+        }
+
+        this.readConnections().wait();
+        this.ftpPublishConnections = _.filter(this.allPublishConnections, c => c.type === PublishService.ALLOWED_CONNECTION_TYPE);
+    }
+
+    public publish(idOrUrl: string, username: string, password: string): IFuture<void> {
+        var ftpConnectionData = this.getFtpConnectionData(idOrUrl, username, password).wait();
+        return this.publishToRemote(ftpConnectionData);
+    }
+
+    public listAllConnections(): void {
+        if (!this.ftpPublishConnections.length) {
+            this.$logger.info("No connections found. To add a connection, run $ appbuilder publish add");
+            return;
+        }
+
+        var table = new Table({
+            head: ["Index", "Name", "Publish URL"],
+            chars: {'mid': '', 'left-mid': '', 'mid-mid': '', 'right-mid': ''}
+        });
+
+        _.forEach(this.ftpPublishConnections, (connection, index) => {
```

Comment:
I don't think _.map is applicable here because within the ._forEach in every iteration I push a small *array* whereas _.map would return an array containing all of the small arrays and I'd still have to iterate it to get its values